src/Classic.jl

```diff
@@ -38,63 +34,15 @@ abstract type Actor{TBehavior} end
 """
 abstract type Addr end
 
-abstract type SendStyle end
-struct Sendable <: SendStyle end
-struct CopySendable <: SendStyle end
-struct Racing <: SendStyle end
-struct NonSendable <: SendStyle end
-
-"""
-    Trait SendStyle: Sendable, CopySendable, Racing, NonSendable
-
-Trait for marking types that can be sent to actors.
-    
-Sending mutable data between actors breaks the model and it may lead to data races,
-so - as a general rule - it should be avoided.
-
-By default, every type is `NonSendable`. You have to mark yor message types as `Sendable`,
-`CopySendable` or `Racing` before sending them. e.g.
-
-```
-abstract type MyMessage end
-Classic.SendStyle(::Type{<:MyMessage}) = Racing()
-
-struct SafeMsg end
-Classic.SendStyle(::Type{SafeMsg}) = Sendable()
-```
-
-`CopySendable` types will be deep copied before sending.
-
-When using `Sendable`, you have to be sure that it is really safe to share the data. Note that
-immutable structs may also contain mutable data deeper down.
-
-With `Racing` you warn the user to trait the message as unsafe shared state. It will
-be delivered to a special `onmessage` method with an extra ::Racing argument for the
-receiver to be warned.
-"""
-SendStyle(::Type{<:Any}) = NonSendable()
-
 """
-    send(sender::Actor, target::Addr, msg)
+    send!(sender::Actor, target::Addr, msg)
 
 Send the message `msg` from `sender` to the actor with `target` address.
-
-The type of `msg` must be marked as `Sendable`, `CopySendable` or `Racing`. See [`SendStyle`](@ref)
 """
-function send end
-
-function send(sender::Actor, target::Addr, msg)
-    send(sender, target, msg, SendStyle(typeof(msg)))
-end
-struct SendingNonSendable <: ActorException
-    type::Type
-end
-send(sender::Actor, target::Addr, msg, ::NonSendable) = throw(SendingNonSendable(typeof(msg)))
-send(sender::Actor, target::Addr, msg, ::CopySendable) = send(sender, target, deepcopy(msg), Sendable())
-# Implementations must at least handle Sendable and Racing
+function send! end
 
 """
-    function spawn(spawner::Actor, behavior)::Addr
+    function spawn(behavior)::Addr
```

Comment:
Ok, then I would suggest to change the docstring to 
```
spawn([context], behavior)::Addr
```
where context can be something, an `Actor`, but also a `Module` (we need this in GenServers).

Reply:
To be sure we are speaking about the same: To allow context not to be written, the `@onmessage` or `@actor` macro is needed. In Actors.jl that macro may not do anything, or only for GenServers, but it has to be written to provide compatibility.  I think that your code will not use the `onmessage` function directly, so this may not be a  bad thing for you, and I can accept it in my code, because in Circo it allows the user not to write out the context. 
Are you OK with that? If yes, I will document the two forms.

Reply:
I think, the first parameter in your description `function spawn(spawner::Actor, behavior)::Addr` is specific to your implementation. Nobody else has it, neither Agha, not Erlang/Elixir, nor Scala ... This should therefore not be in an interface definition referring to the classical model. Why can't you define this method in your library? Why can't you accept my proposal above and then define it further in your library? Or why don't you deliver the context as a keyword ...?

Can't you solve the problem simply with a function barrier in your library, like

```
spawn(behavior) = spawn(me(), behavior)
function spawn(spawner::Actor, behavior)::Addr
     ....
end
```

If somebody works with different actor libraries, she will specify what she spawns with e.g. `Circo.spawn(bhvA)`, `Actors.spawn(bhvB)`, `GenServers.spawn(myGenServerModule)`, `Supervisors.spawn(mySupervisorModule)` and so on. So everybody is safe.

Have you looked at my code? Why do you think, I don't use the `onmessage` function? Or do you speak of an `@onmessage` macro which again is library specific?

Reply:
When I wrote that I think you will not use onmessage directly, I thought that users of Actors.jl will use another interface that is implemented with onmessage, so only a very few place inside Actors.jl needs to use the macro. Sorry if I was not clear and you found it offensive. And I am not sure if this is true, of course.

It is far from being true, that nobody has this. Akka, for example provides the context to the constructor, which the user has to handle. It is then accessible from the self object, which is implicitly available everywhere.
https://developer.lightbend.com/guides/akka-quickstart-java/define-actors.html 

In the Julia universe, exactly the opposite seems to be true: **_You are the only one who has it_**:

https://github.com/JuliaActors/Actors.jl/blob/930472405af0b147a6fc51eaacdb14f29017e21a/src/actor.jl
https://github.com/richiejp/Actors.jl/blob/d3d6c7a0175969d222f21a831e6057d89d21c8a5/src/Actors.j

The whole point of this interface is for the user to _not_ name any implementation in the actor code. Any. Never, except maybe at startup, which can be considered configuration.  E.g. if Circo is added to the project, then Circo will run the actors defined, but if the user wants to replace Circo with a different library, then they simply replace the dependency.

If the user has to name the implementation when an actor spawns another, then the code is not universal. It is like insisting users of the POSIX API to name the operating system.

I hope we can agree that the issue is real.

The context is needed everywhere: in self(), send(), spawn() and become(). Getting it from a global store is a possibility of course, but that it global state (and slow). If the actor is already provided, then the context can be accessed from it, that is why I referenced the creator actor in spawn. We can use a separate context argument, and I would be more happy with it. I just not wanted to introduce a new concept, it seemed easier for you to accept that the actor must be provided.

Please check other implementations to have a better understanding of non-async actor thinking, because it is really hard to discuss technical issues without that. _QuickActors.jl is a good start, it is only 80 lines of trivial code. If you can do it without providing the execution context either explicitly or implicitly through the actor that is running (and without global state), I am happy to go with your solution._

Using the macro would provide more freedom generally, because different implementations can transform the code differently. I think it will be helpful with higher level . I am more and more in support of it. So my current proposal is like this:

```
@actor function onmessage(me::MyActor1, msg::Msg1)
    spawn(AnotherActor())
end
```

I do not like macros, there is not a single one defined yet in Circo, but I am sure we will find this macro useful later.

ps.: I try my best to be constructive, really. If you feel otherwise, I am very sorry! And I am sure you are trying to do the same. Our views are very far from each other, which makes this discussion hard, but this difference is exactly the reason for the resulting common ground to hold a much bigger universe.

Reply:
Let's clarify first what we are (or should be) talking about.

**We are not** talking about

- _our_ context, which may or may not put us into trouble (though it probably does 😉),
- specific implementations (we should avoid that).

**We are** talking about

- the _actor_ context, about which we have less disagreement than you may think,
- the primitive for actor creation, that is `spawn`,
- a docstring in an abstract definition of it in `ActorInterfaces`, which tells implementers what they should implement for
- the classical Actor Model.

## About actor context

In "Creating Actors" on p. 30 Agha doesn't talk about actor context but figure 3.1 (which is an event or a space-time diagram) suggests that a newly created actor inherits a context from its creator due to causality.

Vaughn Vernon in "Reactive Messaging Patterns with the Actor Model" describes on p. 49 the actor creation in Scala/Akka as

> `actorOf(...)` This creates the new actor as a child of this context with the given name ...

and on p. 52

> Every actor has an `ActorContext`, which is accessible through method `context()` on the `Actor`.

So far so good, I agree with everything here. Therefore I propose (as I did initially) to settle our discourse with the following description in the docstring:

```
    spawn(context, behavior)::Addr
    spawn(behavior)::Addr
```

In the second form the newly created actor gets its context implicitly.

## What I'm against

I am against describing it in a common interface as `spawn(spawner::Actor, behavior)::Addr` since it looks like the actor is sharing its actor variable, which it must never do (I hope, we agree here!). I was referring to that when I said that nobody does it. 

## On our debate

I was getting sad about it, since I felt I was getting a fight at every corner even when it was absolutely unnecessary. If we agree on the first points (what we are and are not talking about) we should not get into heated debates and misunderstandings. 

Still I feel that we are making progress and I am learning a lot since it forces me to think through the matter more deeply. And I am not giving up so easily even if I imagined that we could progress more quickly.

Reply:
On our debate: I have the same feeling, and I think it is because you do not understand my points. I also think, that you also think the same. :) And yes, there is already huge progress! We just have to discuss further. Let's do it!

What you are against: I accept that. It was the first proposal, a failed attempt for a compromise, because I had known that in Actors.jl there is no need for explicit context handling.  I just tried to hide the context in the actor. That's it. If you do not like it, no problem, I do not like it either. It was there exactly for you to be  happy. Really. Let's remove that! But the context needs to be handled. (And I already have written this, so here I again have the feeling that you do not understand my points. Maybe I am not clear enough.)

About actor context: What does "gets its context implicitly" means? Please implement it in QuickActors.jl to show me, how it is possible, because I think it is not possible without creating a global store. Do you really insist that implementations create a global store instead of using a context?

Circo, just like the other two Julia actor systems we know of, forces the user to explicitly handle the context everywhere. Please check this line: https://github.com/Circo-dev/CircoCore.jl/blob/c4188b3828f1f66fc3403167df08d2c8cefd8642/src/service.jl#L127 I think it is very clear, what happens here, and yes, explicitly passing the context is a pain, but it can be hidden using a macro. That is what I am
proposing now.

So please also comment on the `@actor` macro solution, my current proposal, which I think would solve the problem! I also think that it will be helpful in the future, e.g with GenServers. Additionally, it opens up the way to simplify the `Actor{Bhv}` definition, which you do not like. That construct was there mostly for providing flexibility for the implementations that can also be provided with the macro. Please check QuickActors.jl to see how the context is stored in the actor itself (just like in Akka). In the current version the onmessage call must be dispatched on `::Actor{ConcreteBhv}`, because the actor stores the context. With the macro, the onmessage call can be dispatched simply on `::ConcreteBhv`, and `self()`, `spawn(otherbhv)`, etc just works, because the macro injects an extra context argument. This is a form of implicit context that works without a global store. The awkward `behavior(me)` can also be eliminated, because `onmessage` is dispatched on the behavior, not on the actor.

I will create a new proposal based on this fork, extended with the macro-based interface. I hope I will get back with it in a few hours.

Reply:
> you do not understand my points. I also think, that you also think the same.

You are right. I still don't understand why we cannot settle this with my proposed description. 

Implicit context means in Julia's `Task(somefunction)` that it inherits a world age and the global variables. I know that the latter is dangerous in concurrent programming. So - as in the Julia manual - I warn the user to avoid them. I learn in this discussion that I perhaps need a section about context in the docs.

If you need an explicit context, just go with the first description, why not? The interface does not prescribe both implementations in each library. Maybe, it would be sufficient to make that clear. Actor creation is deliberate in the libraries. Actors can be created in different ways and with different commands. Call it differently if you like. Each library must have a creation function. It is just convenience to call it `spawn`. Yet less we must prescribe in an elementary interface what parameters we use in this method. Our goal is to make actors communicate after they are created. Therefore the creation function must return an `Addr`. That's enough!

Why should I implement anything in `QuickActors` if it agrees perfectly with the interface? Do I have to demonstrate anything here? I have not suggested that you don't need explicit context.

Why open again the discussion on `Actor{Bhv}`? It doesn't matter if I like it or not. I thought we had agreed on it.

My feeling is that everything becomes ever more complicated, even with introducing macros into the interface. Do we really need them? But I will check it out. Thank you for your advice on that! Let's breathe slowly and deeply 🙂.

Reply:
"The interface does not prescribe both implementations in each library."

It does. A library can only say it implements an interface if every program written against the interface will run. If the interface allows two different method to do the same thing, then the user is free to use any of them. If we still want the user to be able to switch implementations, then all implementations _MUST_ support every method.

This is the main source of our disagreement. I have written about this yesterday.

The context is needed for every actor API call, not just for spawning. I have written this yesterday too.

Please, try to understand why this is true. If you disagree, then I do not see how we will find a common ground. The common ground is not the union but the intersection of implementations.

Reply:
> If we still want the user to be able to switch implementations ...

Great, at least now we have found the difference in our assumptions about an interface. I understand now what made you care so much about context and - hopefully - the following makes you understand why I couldn't.

1. From our first chat I had understood your idea to use `YAActL`'s API with `Circo` working as a backend. That API now gets more ore less integrated into `Actors` and by dispatching `onmessage(A::_ACT, msg)` you can use it. That is available now.
2. Then from your [proposal on Discourse](https://discourse.julialang.org/t/ann-yaactl-0-2-yet-another-actor-library-built-in-julia/48637/30) ("actor-level compatibility between implementations" and "minimalistic interface for the classic model") I had understood that actors from different libraries can communicate. That can be achieved – as demonstrated with `SlowActors` – with a few other steps.
3. Now "switch implementations"! I assume that you mean that a programmer should be able to change her program from `using Actors` to `using Circo` and vice versa and it should run perfectly in both cases. Is that right?

Now stage three seems to be more difficult to achieve and _I need some time_ to digest it and think and work through it. If I  have understood our heated debate correctly you mainly care about context which you must provide "for every actor API call".

That I took "actor-level compatibility between implementations" and "minimalist interface" for stage 2 and you took it for stage 3 was a misunderstanding right from the beginning of the interface discussion.

Still I think that common ground for the first two stages is something. And we should try to reach stage 3.

Reply:
Great!

Yes, I had something like 3. in my mind when I wrote the proposal on Discourse! :) It was more like an idea than a proposal, not elaborated, no surprise that you interpreted differently. I had a strange feeling from the beginning because you wrote it will be easy...
When you have created Actors.jl, I have seen that what you did is different, so with ActorInterfaces.jl I tried to show what I was taking about. In my original post about it ( https://discourse.julialang.org/t/juliaactors-a-road-to-an-actors-ecosystem/49898/9 ) I communicated much cleaner that it has a different goal, but it seems that you were too deep in your world, and I failed to notice it.

Stage 2. is clearly something, it was just not what I wanted, so I was disappointed when I saw it. I will check back again later with fresh eyes! Stage 1, as I wrote earlier, was a premature idea, because after seeing your code, I no more think Circo can make it faster. At least not in its current state.

Back to 3.: Yes, with ActorInterfaces.jl it would be possible to write code that runs on every implementation. This does not mean that different actor libraries are fully interchangeable, it is just the ideal. Libs can have extra features, but when a user writes his code against ActorInterfaces.jl, not naming any implementation, then it will be usable on every system. Possibly an application then mixes actors that `using Actors` with the extra features and actors (from external packages) that only `using ActorInterfaces`. E.g. someone could write a distributed hash table against ActorInterfaces, which can be used both in a webapplication developed on Actors.jl and in a blockchain project running on Circo. Just like POSIX.

Yes, stage 3 is hard. The Classic interface is only the beginning, but hopefully our discussion will be easier from now!

Reply:
> Stage 1, as I wrote earlier, was a premature idea, because after seeing your code, I no more think Circo can make it faster. At least not in its current state.

The slowest part in my code was the execution of a `Func`. But it is down since yesterday from about 270 to  29 ns! The latency now is purely in `Task` creation and in `put! `and `take!` on a `Channel`, which is all pure Julia. It was my basic design decision to go with that. But if those three elements can be done somehow significantly faster, I'm absolutely not bound to them. My code is independent from them. For me it is just convenience to use them.  

Stage 2 is surely something which can yet be improved. But if you look at it, consider that with it you may get a user interface compatible with `Actors`'. If you pull in the `onmessage` interface everything can be done with your actors which can be done with `Actors` ... and everything which will come in the next weeks, error handling, supervisors, monitors, `GenServers` ....

Reply:
As you suggested I wrote a version of `QuickActors` running with my latest stripped down `ActorInterfaces.Classic` and run your example with it: 

```
using ActorInterfaces.Classic

struct Spawner end  # Actor behavior (and empty state)

struct SpawnTree # Command to spawn a tree of actors
    childcount::UInt8
    depth::UInt8
end

function Classic.onmessage(me::Actor{Spawner}, msg::SpawnTree)
    if msg.depth > 0
        for i = 1:msg.childcount
            child = spawn(me.addr.scheduler, Spawner())
            send(child, SpawnTree(msg.childcount, msg.depth - 1))
        end
    end
    return nothing
end

using Test
using QuickActors
import ActorInterfaces.Implementation.Tick # Just a test to see if scheduler lifecycle can also be standardized

const TREE_HEIGHT = 19
const TREE_SIZE = 2^(TREE_HEIGHT + 1) - 1

s = QuickScheduler()
rootaddr = spawn(s, Spawner())
send(rootaddr, SpawnTree(2, TREE_HEIGHT))
println("Building a tree of $TREE_SIZE actors and delivering the same amount of messages")
@time while Tick.tick!(s) end
@test length(s.actorcache) == TREE_SIZE

Building a tree of 1048575 actors and delivering the same amount of messages
  1.144999 seconds (10.51 M allocations: 331.736 MiB)

```

It is somehow slower than your version but it has its context in `task_local_storage`, since I assume that there will be only one scheduler per task. My version of `QuickActors` is in branch "paul".

Reply:
Thanks for working on this!

(There is a `me.addr.scheduler` in spawn, but I see it is working without that)

I think it is reasonable to assume that there is no more than one scheduler per task. 

Interestingly, this version is 2.8x slower on my machine. It should not be so slow, because - after marking the return value of `context()` as  `::QuickScheduler`:

```
julia> @btime QuickActors.context().now
  10.774 ns (0 allocations: 0 bytes)
```
(It was 27ns before)

Also the task-local storage is global state in my view. I know I wrote it is an acceptable solution for task-based actor libs, but I fought for the explicit context exactly because I don't think I would replace the explicit context for a task_local_storage-based implicit but slow one in Circo. I also think I cannot be much faster with my own global storage. The main reason is that type (parameters) of the context (and the scheduler) may differ between threads, or even change during runtime, and cannot be named in the source as a concrete type. So when reading it out from the global store, its type will not be known at compile time.

Global state hurts performance in Julia. And performance is the second most important factor for me, after scalability.

Despite this I can accept this version of the interface if we also provide an alternative that is based on explicit context, and we make clear in the docs that using the implicit one may hurt performance on some implementations.  Of course every implementation must provide both, but in Circo I can do it optionally in with a plugin which has no performance penalty when not loaded.

But I really feel that the explicitly added ctx argument is superior with the macro that injects it automatically.

So please take a look at the macro-based solution! I am still working on the details of it, and I am not sure the macro can deliver the flexibility I hope, but it works for injecting the context as the last parameter of every call, the example looks like this:

```
using ActorInterfaces.Classic

struct Spawner end

struct SpawnTree
    childcount::UInt8
    depth::UInt8
end

@actor function Classic.onmessage(me::Spawner, msg::SpawnTree) # No Actor{} here
    if msg.depth > 0
        for i = 1:msg.childcount
            child = spawn(Spawner())
            send(child, SpawnTree(msg.childcount, msg.depth - 1))
        end
    end
    return nothing
end

using Test
using QuickActors
import ActorInterfaces.Implementation.Tick

const TREE_HEIGHT = 19
const TREE_SIZE = 2^(TREE_HEIGHT + 1) - 1

s = QuickScheduler()
rootaddr = spawn!(s, Spawner())
send!(s, rootaddr, SpawnTree(2, TREE_HEIGHT))
println("Building a tree of $TREE_SIZE actors and delivering the same amount of messages")
@time while Tick.tick!(s) end
@test length(s.actorcache) == TREE_SIZE

Building a tree of 1048575 actors and delivering the same amount of messages
  0.539536 seconds (4.23 M allocations: 171.882 MiB, 23.99% gc time)
```

The `behavior()` call and even the `Actor` type was removed. (behavior was for accessing the actor state, not used in this example but you can check the counter example in QuickActors tests, it looks much better now)  I think this version of the interface is cleaner, despite the macro. Of course `@actor function Classic.onmessage(me::Spawner, msg::SpawnTree)` could look better, e.g. `@onmessage me::Spawner msg::SpawnTree begin ...`, or `@on SpawnTree => Spawner begin ...`, or I don't know.

 You can find the interface in the macro branch, QuickActors is accidentally in its main branch, sorry for that.

Reply:
**First of all** I think we are making progress. I like your solution to inject context (even if I would suggest to name it differently, e.g. `@contx` or `@ctX` or the like for making it more explicit what we are doing here). And I think that I can go with it in general.

**Second** I think that we need yet some time to get it straight. I was thinking through it from a user perspective:

1. Still a user cannot write a mini program like Agha's example for stack or factorial, `spawn` e.g. a stack actor into the context of the reference implementation (`QuickActors`) and then can operate it asynchronously.
2. `onmessage` is an interface too low for a user, I think. It is only a glue between various implementations of `ActorInterfaces`. It cannot replace a user behavior function. And we shouldn't even try to suggest that we create an `@actor` out of it.
3. Then the user must be able to `spawn` an actor into another context (maybe also of another implementation) and the two actors must be able to communicate.

This has various implications: an actor address must provide the context for `send` to work across contexts. An actor itself must have easy access to its context. Maybe we should give the context a bigger weight than behavior and define a actor as one of `Actor{Context}` instead of `Actor{Bhv}` as you mentioned elsewhere (even if this doesn't matter much for a user because the actor is hidden from her). Any implementation of `ActorInterfaces` must provide an asynchronous execution context for users and their programs. ...

4. The interface provides a way to write basic actor programs with the basic primitives. But users will demand higher-level API functions like `receive`, `request`, `call` ... which are not part of the interface. We must give them a way to get those.

Those are only some preliminary thoughts and open ends of this intricate clew and I will work on them maybe next week. For now I will continue to work on `Actors`. I think that we cannot settle all of that so quickly and we will have to put in more work to make it run smoothly.

Reply:
Yes, there is a lot of work to do here, and no need to rush! 

I will rename the macro to `@ctx` to explicitly communicate what it does, and update the docs for the user to see that the macro is just a convenience tool. With the more general name I wanted to allow implementations to do library-specific transformations on the onmessage code, but that would not play nice with package precompilation, so it seems better to drop that idea and only inject the context.

Some thoughts on your points:

1. I will try to implement it to see if something is missing
2. Yes, onmessage is the lowest-level api that is possible. It is fine if some users do not want to use it directly, but semantically it should be enough. To give the user and library developer freedom, higher level apis - especially the ones that add new semantics like concurrency of blocking operations - can be added as extensions. I already created a stub, ActorInterfaces.Armstrong (naming is arbitrary) and I have a few extension ideas, e.g. you deleted SendStyle from Classic, which I respect as it is clearly not related to Agha, but I definitely want to see a solution for the problem it tries to solve, so SendStyle will possibly go into an extension for users to decide if it is helpful or if someone can find a better solution.
3. The context is accessible in the `ctx` argument injected into onmessage, so we can define further functions that operate on it. Cross-spawning seems like an example of an extension for me. 
 4. See 2.

I also have to work on other projects, so let our minds work on this in the background for a while.

Reply:
now in the merged version `Actor` is no longer defined and exported. Is this deliberate?

Reply:
Sorry for the late answer! Yes, Actor is not used anymore so I have deleted it. The user only works with behaviors, the actor itself is now implementation detail, not visible to the actor developer. Do you think this has drawbacks? We can also reintroduce it as an implementation-related type if needed for interoperability between libs.

Reply:
For me that's perfectly fine.

What about registering it? Then I could use it for `Actors` 0.2 which will come in the next 10 days.

Reply:
OK, I will improve the docs a bit and register it soon!

Reply:
I think in the export `onmessage` is yet missing.

Reply:
Intentionally, because the user does not call it, only adds new methods, and that is not possible after simply `using` the 
 interface. To be honest, even after reading several discussions on this, the good exporting strategy it is still a bit unclear to me.

Reply:
ok, yes you are right, its better!

Implementers must reimplement it explicitly as `ActorInterfaces.Classic.onmessage` or if `Classic` is imported as `Classic.onmessage`.

This must go into the docs.